### PR TITLE
Launch Security with AMD SEV

### PR DESCRIPTION
**Pull request by vasiliy-ul**
**What this PR does / why we need it**:

This is a follow-up of the mailing list discussion [0]. The PR introduces new elements in the VMI spec: `launchSecurity` and `sev`. By setting those the user can enable guest memory encryption using SEV (Secure Encrypted Virtualization) extension available on AMD's EPYC CPUs [1].

SEV technology has been around for quite some time already and its supported by recent qemu and libvirt [2][3]. There are new extensions like SEV-ES (Encrypted State) and SEV-SNP (Secure Nested Paging) which bring more features. Though this PR focuses on the basic SEV.

[0] https://groups.google.com/g/kubevirt-dev/c/a6XlT0qRNAY/m/Hex2r3lDBwAJ
[1] https://developer.amd.com/sev/
[2] https://libvirt.org/kbase/launch_security_sev.html
[3] https://libvirt.org/formatdomain.html#launch-security

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

The new elements in the VMI spec are currently mapped to what is expected in libvirt domxml. They look like the following:

```yaml
spec:
  domain:
    launchSecurity:
      sev: {}
```

In order to enable SEV all `virtio devices` need to be configured with the `iommu='on'` attribute in their `<driver>` configuration. In addition, all memory regions used by the VM must be locked for Direct Memory Access (DMA) and to prevent swapping.

Known limitations (some of them will probably be removed in the future):
- Requires AMD EPYC CPU (!)
- SEV is only going to work with OVMF (UEFI)
- SEV-encrypted VMs cannot contain directly-accessible host devices (that is, PCI passthrough)
- Live Migration is not supported

**Release note**:

```release-note
Experimental support of AMD SEV
```

### CODE REVIEW

**Comment by fabiand** [comment]
@vladikr who can help to review this pr? Maybe somebody from the capybara team?

File: pkg/virt-launcher/virtwrap/launchsecurity/virsh.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+func (v *virsh) Domcapabilities() ([]byte, error) {
+	return exec.Command("virsh",
+		"domcapabilities",
+		"--machine", "q35",
+		"--arch", "x86_64",
+		"--virttype", "kvm",
+	).Output()
+}
```
Comment (vladikr) [inline pkg/virt-launcher/virtwrap/launchsecurity/virsh.go:50]: I think this is not needed as we can already obtain the capabilities in the virt-launcher.

**Review by vladikr — CHANGES_REQUESTED** [review]
@vasiliy-ul Thank you for this PR.
I've made some comments.

My main concern is the current API. What I don't understand is how the attestation is coming into play here? Perhaps you can elaborate on this?

File: staging/src/kubevirt.io/client-go/api/v1/schema.go [hunk 2]
```diff
@@ -631,6 +634,44 @@ type DiskTarget struct {
...
+	SEV *SEV `json:"sev,omitempty"`
+}
+
+//
+// +k8s:openapi-gen=true
+type SEV struct {
+	// CBit Position
+	Cbitpos *uint `json:"cbitpos,omitempty"`
+	// No. of physical bits in address space
+	ReducedPhysBits *uint `json:"reducedPhysBits,omitempty"`
```
[the remark this answers is not in the document]
Comment (dagrh) [reply staging/src/kubevirt.io/client-go/api/v1/schema.go:650]: I suggest omitting the cbitpos and reducedPhysBits; current qemu (and libvirt?) automatically figure them out, and they're different on newer hardware (Milan)

**Comment by dagrh** [comment]
I like that you have names for the policy elements, rather than just relying on the awkward bit mask.
(Which I think is probably also useful in SNP where I think the policy gets more complex).
[I wonder if you need something more abstract as well though (depending on how smart images get) - could we make something where we can ask for confidential computing and it would get the best thing supported by the host; TDX on Intel, SEV.. on AMD? -  anyway that's a more abstract thought].

**Comment by vasiliy-ul** [comment]
Hey, @vladikr , @dagrh , thank you for the review comments.

> My main concern is the current API. What I don't understand is how the attestation is coming into play here? Perhaps you can elaborate on this?

Initially my idea was as the first step to bring the basic support of SEV memory encryption and address the attestation at a later point. Currently afaik there are still some missing pieces in e.g. libvirt for that. Also for the attestation it would probably be nice to consider various use-cases (SEV, SNP, TDX, ...) in order to have a unified user-facing API. So this is transformig into a bigger topic touching not only the basic SEV. Do you think it makes sense to include attestation right from the beginning?

> [I wonder if you need something more abstract as well though (depending on how smart images get) - could we make something where we can ask for confidential computing and it would get the best thing supported by the host; TDX on Intel, SEV.. on AMD? - anyway that's a more abstract thought].

That's a good idea. E.g. if I dont care I can specify smth like:

```yaml
spec:
  domain:
    launchSecurity: {}
```
or be more precise to select only SEV:
```yaml
spec:
  domain:
    launchSecurity:
      sev: {}
```
The VM can then be scheduled on a node with the corresponding capabilities.

File: staging/src/kubevirt.io/client-go/api/v1/schema.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (vasiliy-ul) [reply staging/src/kubevirt.io/client-go/api/v1/schema.go:650]: right, I have a note in the PR description about it:

>NOTE: in current implementation there is an automatic detection of cbitpos and reducedPhysBits. That makes it redundant to have them in the VMI spec and potentially they can be removed. 

I was thinking about potential benefit of having those settings in the spec. At the end just decided to make them optional. So you both suggest to remove them completly instead, right?

Also what do you think about the policy here? Currently the flags are passed as list of strings. Would it be better to have smth more explicit e.g.


```yaml
spec:
  domain:
    launchSecurity:
      sev:
        policy:
          encryptedState: true
```
vs
```yaml
spec:
  domain:
    launchSecurity:
      sev:
        policy:
          - EncryptedState
```

File: pkg/virt-launcher/virtwrap/launchsecurity/virsh.go (hunk 1 above)
Comment (vasiliy-ul) [reply pkg/virt-launcher/virtwrap/launchsecurity/virsh.go:50]: Do you mean it is better to pass the capabilities via `VirtualMachineOptions` from `virt-handler`? Initially I was looking at that option. But in practice that would require modifications of the node-labeler (since currently it does not save the complete  domcapabilities AFAIK) and some additional logic to send and recieve that data during VMI sync.... just calling `virsh domcapabilities` from `virt-launcher` turned out to be simpler and I think it should not cause issues.

Potentially at some point it would be usefull to also label a node as SEV-capable. Though at the moment I have not included that part here (it can be introduced in a follow-up PR) hence decided to go with the simple approach at first. Do you think it will be more beneficial to send the data from `virt-handler` at current step?

File: pkg/virt-controller/services/template.go [hunk 3]
```diff
@@ -1078,6 +1079,12 @@ func (t *templateService) renderLaunchManifest(vmi *v1.VirtualMachineInstance, t
 		}
 	}
 
+	if util.IsSEVVMI(vmi) {
+		// privileged is required for memory locking
+		privileged = true
```
[the remark this answers is not in the document]
Comment (vasiliy-ul) [reply pkg/virt-controller/services/template.go]: Good point. Initially I missed that there was memory locking already implemented for vfio.

File: pkg/virt-launcher/virtwrap/launchsecurity/virsh.go (hunk 1 above)
Comment (vasiliy-ul) [reply pkg/virt-launcher/virtwrap/launchsecurity/virsh.go]: UPD: I think my prev comment is outdated. After looking at it one more time I've updated the code and now the SEV configuration is obtained by node-labeller and virt-launcher gets it from virt-hanldler.

**Comment by vasiliy-ul** [comment]
This should be ready for another review round. @vladikr , maybe you can take one more look when have time? I think all the comments are addressed. Thanks.

File: pkg/virt-launcher/virtwrap/launchsecurity/virsh.go (hunk 1 above)
Comment (vasiliy-ul) [reply pkg/virt-launcher/virtwrap/launchsecurity/virsh.go]: One more update: I just realized that KubeVirt has recently switched to a newer libvirt (7.6.0). That makes the whole story much easier since we can leverage libvirt to automatically fill `cbitpos` and `reducedphysbits`. So no need to query domcapabilities anymore.

**Comment by vasiliy-ul** [comment]
Some updates: I had to bump the EDK2 rpm here since the original one did not ship the roms with SEV support. Also made node-labeller aware of SEV. The VMs are now scheduled on the nodes with the corresponding label. Still the PR is ready for review. Probably there is some room for discussion around the new VMI spec APIs. PTAL.

Ping @dagrh , @vladikr , @rmohr , @davidvossel . Would appreciate your opinions/thoughts about the PR. Thanks!

File: pkg/virt-handler/isolation/detector.go [hunk 4]
```diff
@@ -119,7 +119,7 @@ func (s *socketBasedIsolationDetector) Allowlist(controller []string) PodIsolati
 
 func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInstance) error {
 	// only VFIO attached or with lock guest memory domains require MEMLOCK adjustment
-	if !util.IsVFIOVMI(vm) && !vm.IsRealtimeEnabled() {
+	if !util.IsVFIOVMI(vm) && !vm.IsRealtimeEnabled() && !util.IsSEVVMI(vm) {
```
Comment (vladikr) [inline pkg/virt-handler/isolation/detector.go:122]: Out of curiosity, does the use of a secure launch with QEMU generate an additional memory overhead?

**Review by vladikr — COMMENTED** [review]
Overall looks great, thank you @vasiliy-ul  for all of this great work!
I only have a few minor questions.

File: tests/launchsecurity/sev.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+		By("Expecting the VirtualMachineInstance console")
+		Expect(console.LoginToFedora(vmi)).To(Succeed())
+
+		By("Verifying that SEV is enabled in the guest")
+		err := console.SafeExpectBatch(vmi, []expect.Batcher{
+			&expect.BSnd{S: "\n"},
+			&expect.BExp{R: console.PromptExpression},
+			&expect.BSnd{S: "dmesg | grep --color=never SEV\n"},
+			&expect.BExp{R: "AMD Secure Encrypted Virtualization \\(SEV\\) active"},
```
Comment (jean-edouard) [inline tests/launchsecurity/sev.go:34]: Nit: using a raw string (backticks) should avoid having to escape the escape character (backslash)

**Review by jean-edouard — COMMENTED** [review]
Really good stuff, thank you!
Just a few minor questions/comments in-line.

File: staging/src/kubevirt.io/api/core/v1/schema.go [hunk 6]
```diff
@@ -598,6 +601,40 @@ type DiskTarget struct {
...
+	SEV *SEV `json:"sev,omitempty"`
+}
+
+type SEV struct {
+	// Guest policy flags as defined in AMD SEV API specification.
+	// Note: due to security reasons it is not allowed to enable guest debugging. Therefore NoDebug flag is not exposed to users and is always true.
+	Policy *SEVPolicy `json:"policy,omitempty"`
+}
+
+type SEVPolicy struct {
```
[the remark this answers is not in the document]
Comment (vasiliy-ul) [reply staging/src/kubevirt.io/api/core/v1/schema.go:615]: Well, mostly in theory :) Unfortunately the documentaiton is not very 'detailed'. From what I know:

- `NoDebug` - if set to `true` prevents debugging the guest memory with SEV ioctl cmds `KVM_SEV_DBG_DECRYPT`/`KVM_SEV_DBG_ENCRYPT`. In this PR it is not exposed in VMI spec and always enabled due to security reasons.
- `EncryptedState` - controls whether `SEV ES` should be enabled or not. This PR does not introduce 'full' support of the feature. It only allows setting this policy flags. There is still a need to detect `ES` capability on the node and label it appropriately. I will work on it in the follow-up PR.
- other policy flags AFAIK control the behaviour during snapshot/restore and migration flows (which are currently not supported)

One more thing is that the policy bitmask is used during the attestation and it is included in the measuremnt. So once it is set by a user it cannot be changed during the lifetime of the guest.

File: pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go [hunk 7]
```diff
@@ -978,4 +978,25 @@ var _ = Describe("VirtualMachineInstance Mutator", func() {
...
+
+	It("should add SEV node label selector with SEV workload", func() {
+		vmi.Spec.Domain.LaunchSecurity = &v1.LaunchSecurity{SEV: &v1.SEV{}}
+		vmiSpec, _ := getVMISpecMetaFromResponse()
+		Expect(vmiSpec.NodeSelector).NotTo(BeNil())
+		Expect(vmiSpec.NodeSelector).To(BeEquivalentTo(map[string]string{v1.SEVLabel: ""}))
+	})
+
+	It("should not add SEV node label selector when no SEV workload", func() {
+		vmi.Spec.Domain.LaunchSecurity = &v1.LaunchSecurity{}
```
[the remark this answers is not in the document]
Comment (vasiliy-ul) [reply pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator_test.go:990]: Currently an empty `LaunchSecurity` setting in the VMI spec is simply ignored. I personally like the idea suggested by @dagrh  when a user requests `vmi.Spec.Domain.LaunchSecurity = &v1.LaunchSecurity{}` and KubeVirt automatically schedules the VM depending on the capabilities of the available nodes (which can be SEV, SEV ES, TDX, etc.). Though I decided to leave it for the follow-up PR when at least SEV ES will be supported (to have smth to choose from).

File: pkg/virt-launcher/virtwrap/converter/network.go [hunk 8]
```diff
@@ -120,6 +120,18 @@ func createDomainInterfaces(vmi *v1.VirtualMachineInstance, domain *api.Domain,
...
+			if domainIface.Rom != nil && domainIface.Rom.Enabled != "no" {
+				return nil, fmt.Errorf("%s: Rom.Enabled = '%s' not allowed with launch security", iface.Name, domainIface.Rom.Enabled)
+			}
+			domainIface.Rom = &api.Rom{Enabled: "no"}
+			if domainIface.Driver != nil {
+				domainIface.Driver.IOMMU = "on"
+			} else {
+				domainIface.Driver = &api.InterfaceDriver{Name: "vhost", IOMMU: "on"}
+			}
+		}
```
[the remark this answers is not in the document]
Comment (vasiliy-ul) [reply pkg/virt-launcher/virtwrap/converter/network.go:134]: TBH I was not able to find the flow when `Rom.Enabled` is set to `yes`. So this is mostly a precaution measure for the case when someone changes the network configuration code. Besides, the `Rom` element is not exposed in the VMI spec and moving the check to the validation webhook will require similar logic as in the converter. Does it seem reasonable?

File: pkg/virt-handler/isolation/detector.go (hunk 4 above)
Comment (vasiliy-ul) [reply pkg/virt-handler/isolation/detector.go:122]: Indeed libvirt documentaiton suggests to add `256MiB` extra memory. Seems that I forgot about it at some point. Thanks, good catch :+1:

File: pkg/virt-launcher/virtwrap/converter/converter.go [hunk 9]
```diff
@@ -1241,6 +1258,21 @@ func Convert_v1_VirtualMachineInstance_To_api_Domain(vmi *v1.VirtualMachineInsta
 		}
 
 	}
+	// Set SEV launch security parameters: https://libvirt.org/formatdomain.html#launch-security
+	if c.UseLaunchSecurity {
+		if !vmi.IsBootloaderEFI() {
```
[the remark this answers is not in the document]
Comment (vasiliy-ul) [reply pkg/virt-launcher/virtwrap/converter/converter.go:1263]: Right, I think I decided to keep it as a precaution measure. Just to have a meaningfull error message in case the admition code is reworked at some point (e.g. if someone accidentally changes the validation logic). Do you think it would be better to drop it from here?

File: tests/containerdisk/containerdisk.go [hunk 10]
```diff
@@ -36,6 +36,7 @@ const (
 	ContainerDiskVirtio               ContainerDisk = "virtio-container-disk"
 	ContainerDiskEmpty                ContainerDisk = "empty"
 	ContainerDiskFedoraRealtime       ContainerDisk = "fedora-realtime"
+	ContainerDiskFedoraUefi           ContainerDisk = "fedora-uefi"
```
[the remark this answers is not in the document]
Comment (vasiliy-ul) [reply tests/containerdisk/containerdisk.go:39]: Right, I haven't yet updated this part. I tested with Fedora 35 from https://github.com/kubevirt/kubevirtci/pull/713 and it works fine. Will switch to that image.

File: pkg/virt-launcher/virtwrap/launchsecurity/sev_test.go [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+	"k8s.io/utils/pointer"
+
+	v1 "kubevirt.io/api/core/v1"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/launchsecurity"
+)
+
+var _ = Describe("LaunchSecurity: AMD Secure Encrypted Virtualization (SEV)", func() {
+	Context("SEV policy conversion", func() {
+		const NoDebugBit uint = 0b1
```
[the remark this answers is not in the document]
Comment (vasiliy-ul) [reply pkg/virt-launcher/virtwrap/launchsecurity/sev_test.go]: updated

File: pkg/virt-launcher/virtwrap/launchsecurity/sev_test.go [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+			*field = pointer.BoolPtr(true)
+			Expect(launchsecurity.SEVPolicyToBits(&policy)).To(Equal(NoDebugBit | expectedBit))
+			*field = pointer.BoolPtr(false)
+			Expect(launchsecurity.SEVPolicyToBits(&policy)).To(Equal(NoDebugBit))
+		},
+			table.Entry("NoKeysSharing ", uint(0b000010), &policy.NoKeysSharing),
+			table.Entry("EncryptedState", uint(0b000100), &policy.EncryptedState),
+			table.Entry("NoSend        ", uint(0b001000), &policy.NoSend),
+			table.Entry("Domain        ", uint(0b010000), &policy.Domain),
+			table.Entry("SEV           ", uint(0b100000), &policy.SEV),
```
[the remark this answers is not in the document]
Comment (vasiliy-ul) [reply pkg/virt-launcher/virtwrap/launchsecurity/sev_test.go]: updated

File: pkg/virt-launcher/virtwrap/efi/efi_test.go [hunk 13]
```diff
@@ -52,16 +52,18 @@ var _ = Describe("EFI environment detection", func() {
 			efiEnv := DetectEFIEnvironment(arch, ovmfPath)
 			Expect(efiEnv).ToNot(BeNil())
 
-			Expect(efiEnv.Bootable(true)).To(Equal(SBBootable))
-			Expect(efiEnv.Bootable(false)).To(Equal(NoSBBootable))
+			Expect(efiEnv.Bootable(true, false)).To(Equal(SBBootable))
```
[the remark this answers is not in the document]
Comment (vasiliy-ul) [reply pkg/virt-launcher/virtwrap/efi/efi_test.go]: Indeed with those true/false it was hard to follow the logic. In general no strong objections about the struct. But defined the bool constants simply because of personal preference. Now it looks much better. Thanks.

File: tests/launchsecurity/sev.go (hunk 5 above)
Comment (vasiliy-ul) [reply tests/launchsecurity/sev.go]: I actually spent quite some time figuring out that my test was failing because of non-escaped backslash here :) Anyway, updated as suggested, maybe raw string will improve readability here.

File: pkg/virt-launcher/virtwrap/converter/network.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (jean-edouard) [reply pkg/virt-launcher/virtwrap/converter/network.go:134]: This happens to me regularly (not finding the flow). Usually, it's because we just pass the value to libvirt, where the logic actually is.
It's the case here. See https://github.com/kubevirt/kubevirt/blob/403f1a864e58cc5db9774bf7fe5880e281d3d638/pkg/virt-launcher/virtwrap/api/schema.go#L708
Rom defaults to yes in libvirt, we just set it to no when the user didn't define a boot order for the interface, to disable iPXE.
Your code here probably needs reworking to take into account that the Rom attribute is true by default.

File: pkg/virt-launcher/virtwrap/converter/converter.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (jean-edouard) [reply pkg/virt-launcher/virtwrap/converter/converter.go:1263]: I'd say it's fine to leave it, but the error message could be a bit more dramatic! Like `/!\ This should never happen!! /!\ A VMI with SEV + BIOS was admitted!!` :)

File: pkg/virt-launcher/virtwrap/converter/network.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (vasiliy-ul) [reply pkg/virt-launcher/virtwrap/converter/network.go:134]: Yeah, I somehow did not take into account the defaults from libvirt. Valid point. Will look at how to rework the code here. @jean-edouard , thanks for the info :+1:

File: tests/launchsecurity/sev.go (hunk 5 above)
Comment (vasiliy-ul) [reply tests/launchsecurity/sev.go]: ... and the funny thing is that in newer kernels this message has been changed to "AMD Memory Encryption Features active: SEV". So no need to escape anything when switching to Fedora35 :)

Comment (jean-edouard) [reply tests/launchsecurity/sev.go]: Urg.. I assume there's no sys node or whatever that we could check instead of dmesg?

Comment (vasiliy-ul) [reply tests/launchsecurity/sev.go]: AFAIK no.. not from the guest :( But I will double-check: previously I tested with an older kernel and Fedora 35 uses a newer one, perhaps smth has changed... but I would not be so optimistic about it.

File: tests/containerdisk/containerdisk.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (vasiliy-ul) [reply tests/containerdisk/containerdisk.go]: updated. now its ready for fedora 35.

File: pkg/virt-launcher/virtwrap/converter/converter.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (vasiliy-ul) [reply pkg/virt-launcher/virtwrap/converter/converter.go]: actually after thinking a bit more about it decided not to be that paranoid and dropped the check :)

File: pkg/virt-launcher/virtwrap/converter/network.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (vasiliy-ul) [reply pkg/virt-launcher/virtwrap/converter/network.go]: Updated the code to handle the defaults. Now Rom is explicitly disabled if SEV is requested.

File: tests/launchsecurity/sev.go (hunk 5 above)
Comment (vasiliy-ul) [reply tests/launchsecurity/sev.go]: The kernel checks if SEV is active by reading the MSRs and verifying the proper bits. Unfortunately apart from that dmesg message I cannot see anything that may point to current feature status.

**Comment by vasiliy-ul** [comment]
@vladikr ,  @jean-edouard , I think all the comments are addressed. Please let me know if I missed something.

File: pkg/virt-launcher/virtwrap/converter/network.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (jean-edouard) [reply pkg/virt-launcher/virtwrap/converter/network.go]: I think that approach is fine, but only if no `bootOrder` was specified for the interface, and if `rom` was not manually enabled.
Otherwise, that would mean we silently override parameters explicitly set by the user, which doesn't sound good...
I'd suggest keeping this code as is, but in the admitter just reject VM(I)s that request bootable NICs + SEV.
How does that sound?

**Comment by jean-edouard** [comment]
> @vladikr , @jean-edouard , I think all the comments are addressed. Please let me know if I missed something.

Almost there! I have one last concern on the NIC Rom stuff, please see the corresponding thread.

File: pkg/virt-launcher/virtwrap/converter/network.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (vasiliy-ul) [reply pkg/virt-launcher/virtwrap/converter/network.go]: I was rather thinking about documenting that with SEV Rom will be disabled. But maybe you are right: having a validation in the admitter might be a better option.

File: pkg/virt-controller/services/template.go [hunk 14]
```diff
@@ -1903,6 +1908,12 @@ func GetMemoryOverhead(vmi *v1.VirtualMachineInstance, cpuArch string) *resource
 
 	addProbeOverheads(vmi, overhead)
 
+	// Consider memory overhead for SEV guests.
+	// Additional information can be found here: https://libvirt.org/kbase/launch_security_sev.html#memory
+	if util.IsSEVVMI(vmi) {
+		overhead.Add(resource.MustParse("256Mi"))
```
Comment (vladikr) [inline pkg/virt-controller/services/template.go:1914]: :+1:

File: staging/src/kubevirt.io/api/core/v1/schema.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (vladikr) [reply staging/src/kubevirt.io/api/core/v1/schema.go:615]: Thank you for the explanation. Perhaps we didn't need to expose the `EncryptedState` right now or at least it would be good to have a webhook that rejects VMI specs that uses it. 
I'm not very worried about this as this feature is behind a feature gate and this PR is very large already. Perhaps you can follow up?

[the remark this answers is not in the document]
Comment (vasiliy-ul) [reply staging/src/kubevirt.io/api/core/v1/schema.go:615]: Sure, I also plan to bring in SEV ES and SNP support. Agree that this PR is already big enough. As soon as the basic SEV gets in I will work on those features (+attestation) in the follow-up PRs.

**Comment by vasiliy-ul** [comment]
Rebased the PR. No new code changes.

File: staging/src/kubevirt.io/api/core/v1/schema.go [hunk 15]
```diff
@@ -598,6 +601,40 @@ type DiskTarget struct {
...
+	// +optional
+	NoSend *bool `json:"noSend,omitempty"`
+	// The guest must not be transmitted to another platform that is not in the domain.
+	// Defaults to false.
+	// +optional
+	Domain *bool `json:"domain,omitempty"`
+	// The guest must not be transmitted to another platform that is not SEV capable.
+	// Defaults to false.
+	// +optional
+	SEV *bool `json:"sev,omitempty"`
```
Comment (rmohr) [inline staging/src/kubevirt.io/api/core/v1/schema.go:635]: All these knobs look a little bit too detailed for the initial stab for me. @vasiliy-ul do you think you can trim it down to the absolut minimum right now? 

Or maybe they make sense, but I am not sure what to do with bools like `domain` or `sev`. I would expect by default that if sev is enabled, that it i encrypted and that is basically it. Is "another" platform another node? If yes, this is probably related, but it may not be the right exposure granularity.

**Review by rmohr — COMMENTED** [review]
@vasiliy-ul could you update the PR description? I may be wrong, but it looks like the PR evolved quite a bit since the PR description was last modified :)

File: staging/src/kubevirt.io/api/core/v1/schema.go (hunk 15 above)
Comment (vasiliy-ul) [reply staging/src/kubevirt.io/api/core/v1/schema.go:635]: Then probably I will completely remove the policy flags from the VMI spec... at least for now they are not very usefull. They can be added later gradually along with the corresponding features.

Comment (rmohr) [reply staging/src/kubevirt.io/api/core/v1/schema.go:635]: > Then probably I will completely remove the policy flags from the VMI spec... at least for now they are not very usefull. They can be added later gradually along with the corresponding features.

That sounds great. Thanks!

Comment (vasiliy-ul) [reply staging/src/kubevirt.io/api/core/v1/schema.go]: Done. Removed the policy in a separate commit so if needed the logic can be restored from the git history.

**Comment by vasiliy-ul** [comment]
> @vasiliy-ul could you update the PR description? I may be wrong, but it looks like the PR evolved quite a bit since the PR description was last modified :)

Yep, the escription was a bit outdated. Updated.

**Comment by vladikr** [comment]
Thanks @vasiliy-ul 
Looks good to me now.
/lgtm

**Comment by vladikr** [comment]
uhh I didn't notice it was approved already.
Let's wait for @rmohr to have another look.

File: pkg/virt-config/feature-gates.go [hunk 16]
```diff
@@ -44,6 +44,7 @@ const (
 	DownwardMetricsFeatureGate = "DownwardMetrics"
 	NonRoot                    = "NonRootExperimental"
 	ClusterProfiler            = "ClusterProfiler"
+	LaunchSecurityGate         = "LaunchSecurity"
```
Comment (rmohr) [inline pkg/virt-config/feature-gates.go:47]: Can we call this `LaunchSecuritySEV`? I don't want to spend too much time on this, but maybe `WorkloadEncryptionSEV`?

File: rpm/BUILD.bazel [hunk 17]
```diff
@@ -721,6 +721,8 @@ rpmtree(
         "/var/run": "../run",
         "/usr/sbin/iptables": "/usr/sbin/iptables-legacy",
         "/usr/bin/nc": "/usr/bin/ncat",
+        # Create a symlink to OVMF binary with SEV support (edk2 rpm does not do that for unknown reason)
+        "/usr/share/OVMF/OVMF_CODE.cc.fd": "../edk2/ovmf/OVMF_CODE.cc.fd",
```
Comment (rmohr) [inline rpm/BUILD.bazel:725]: It probably does it in a post-install task which we don't run. :)

**Review by rmohr — APPROVED** [review]
/lgtm

One nit about the feature gate name which I want us to think about.

File: pkg/virt-config/feature-gates.go (hunk 16 above)
Comment (vasiliy-ul) [reply pkg/virt-config/feature-gates.go:47]: Initially I called it `LaunchSecurity` just because it maps to the libvirt naming in domxml. But yeah, I can rename it. `WorkloadEncryptionSEV` seems also good for me.

Comment (rmohr) [reply pkg/virt-config/feature-gates.go:47]: Awesome :+1:

**Comment by vasiliy-ul** [comment]
Rebased the PR. Apart from renaming the feature gate there are no code changes.

File: rpm/BUILD.bazel (hunk 17 above)
Comment (vasiliy-ul) [reply rpm/BUILD.bazel:725]: But it does symlink other OVMF binaries...

**Comment by vasiliy-ul** [comment]
Is it okay to unhold the PR? @vladikr , @rmohr ?

**Comment by rmohr** [comment]
> Is it okay to unhold the PR? @vladikr , @rmohr ?

/unhold

Yes :)
